package/MDAnalysis/coordinates/timestep.pyx

```diff
@@ -0,0 +1,982 @@
+          when attributes require trajectory manipulation (e.g. dt)
+        dt : float (optional)
+          The time difference between frames (ps).  If :attr:`time`
+          is set, then `dt` will be ignored.
+        time_offset : float (optional)
+          The starting time from which to calculate time (in ps)
+
+
+        .. versionchanged:: 0.11.0
+           Added keywords for `positions`, `velocities` and `forces`.
+           Can add and remove position/velocity/force information by using
+           the ``has_*`` attribute.
+        .. versionchanged:: 2.3.0
+           Added the `dtype` kwarg to specify a type for the timestep.
+        """
+        # python objects
+        if dtype not in (np.float32, np.float64):
+            raise TypeError("dtype must be one of (np.float32, np.float64)")
+        self._dtype = dtype
+
+        self.data = {}
+
+        for att in ('dt', 'time_offset'):
+            try:
+                self.data[att] = kwargs[att]
+            except KeyError:
+                pass
+        try:
+            # do I have a hook back to the Reader?
+            # can possibly __cinit__ this with PyWeakRefNew
+            self._reader = weakref.ref(kwargs['reader'])
+        except KeyError:
+            pass
+
+        self.has_positions = kwargs.get('positions', True)
+        self.has_velocities = kwargs.get('velocities', False)
+        self.has_forces = kwargs.get('forces', False)
+
+        # set up aux namespace for adding auxiliary data
+        self.aux = Namespace()
+
+    def __dealloc__(self):
+        pass
+
+    @property
+    def n_atoms(self):
+        """A read only view of the number of atoms this Timestep has
+
+        .. versionchanged:: 0.11.0
+           Changed to read only property
+        """
+        # In future could do some magic here to make setting n_atoms
+        # resize the coordinate arrays, but
+        # - not sure if that is ever useful
+        # - not sure how to manage existing data upon extension
+        return self._n_atoms
+
+    @property
+    def dtype(self):
+        """The NumPy dtype of the timestep, all arrays in the timestep will
+            have this dtype (yet to be finalised)
+
+        .. versionadded:: 2.3.0
+           Added dtype
+        """
+        return self._dtype
+
+    @property
+    def has_positions(self):
+        """A boolean of whether this Timestep has position data
+        This can be changed to ``True`` or ``False`` to allocate space for
+        or remove the data.
+
+        .. versionadded:: 0.11.0
+        """
+        return self._has_positions
+
+    @has_positions.setter
+    def has_positions(self, val):
+        if val and not self._has_positions:
+            # Setting this will always reallocate position data
+            # ie
+            # True -> False -> True will wipe data from first True state
+            self._pos = cnp.PyArray_ZEROS(
+                2, self._particle_dependent_dim, self._typenum, 0)
```

Comment:
This is reallocating right?  So if we had a pathological traj which had saved positions every other frame, we'd be reallocating every other frame?  Maybe better could be another if inside, which is if the array exists, memset(0)

Reply:
I have added a flag that tracks allocation of the array and either zero fills using `PyArray_FILLWBYTE (memset under hood)` if the array is already allocd OR allocs it to zero with `PyArray_ZEROS` with the correct shape if its the first time.

This way we can avoid lots of reallocs, without carrying around a bunch of large uninitialised arrays from `__cinit__`